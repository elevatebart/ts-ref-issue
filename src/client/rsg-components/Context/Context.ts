import React from 'react';
import * as Rsg from '../../../typings';

const StyleGuideContext = React.createContext<StyleGuideContextContents>({
	codeRevision: 0,
	cssRevision: '0',
	config: {} as Rsg.ProcessedStyleguidistConfig,
	slots: {},
	displayMode: 'collapse',
});

export default StyleGuideContext;

export interface SlotObject {
	id: string;
	render: React.FunctionComponent<any>;
}

export interface StyleGuideContextContents {
	codeRevision: number;
	cssRevision: string;
	config: Rsg.ProcessedStyleguidistConfig;
	slots: Record<string, (SlotObject | React.FunctionComponent<any>)[]>;
	displayMode: string;
}

export function useStyleGuideContext(): StyleGuideContextContents {
	return React.useContext(StyleGuideContext);
}
