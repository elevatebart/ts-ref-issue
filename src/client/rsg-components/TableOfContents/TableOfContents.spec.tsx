import React from 'react';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import noop from 'lodash/noop';
import TableOfContents from './TableOfContents';
import { TableOfContentsRenderer } from './TableOfContentsRenderer';
import Context from '../Context';

const components = [
	{
		name: 'Button',
		slug: 'button',
	},
	{
		name: 'Input',
		slug: 'input',
	},
	{
		name: 'Textarea',
		slug: 'textarea',
	},
];

const sections = [
	{
		name: 'Introduction',
		slug: 'introduction',
		content: 'intro.md',
	},
	{
		name: 'Buttons',
		slug: 'buttons',
		components: [
			{
				name: 'Button',
				slug: 'button',
			},
		],
	},
	{
		name: 'Forms',
		slug: 'forms',
		components: [
			{
				name: 'Input',
				slug: 'input',
			},
			{
				name: 'Textarea',
				slug: 'textarea',
			},
		],
	},
];

it('should render a renderer', () => {
	const actual = shallow(
		<TableOfContents
			sections={[
				{
					name: 'Input',
					slug: 'input',
					components,
				},
			]}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should render renderer with sections with nested components', () => {
	const actual = shallow(<TableOfContents sections={sections} />);

	expect(actual).toMatchSnapshot();
});

it('should filter list when search field contains a query', () => {
	const searchTerm = 'but';
	const actual = shallow(
		<TableOfContents
			sections={[
				{
					name: 'Input',
					slug: 'input',
					components,
				},
			]}
		/>
	);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('should filter section names', () => {
	const searchTerm = 'frm';
	const actual = shallow(<TableOfContents sections={sections} />);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('renderer should render table of contents', () => {
	const searchTerm = 'foo';
	const actual = shallow(
		<TableOfContentsRenderer classes={{}} searchTerm={searchTerm} onSearchTermChange={noop}>
			<div>foo</div>
		</TableOfContentsRenderer>
	);

	expect(actual).toMatchSnapshot();
});

it('should call a callback when input value changed', () => {
	const onSearchTermChange = jest.fn();
	const searchTerm = 'foo';
	const newSearchTerm = 'bar';
	const actual = shallow(
		<TableOfContentsRenderer
			classes={{}}
			searchTerm={searchTerm}
			onSearchTermChange={onSearchTermChange}
		>
			<div>foo</div>
		</TableOfContentsRenderer>
	);

	actual.find('input').simulate('change', {
		target: {
			value: newSearchTerm,
		},
	});

	expect(onSearchTermChange).toBeCalledWith(newSearchTerm);
});

it('should render content of subsections of a section that has no components', () => {
	const actual = shallow(
		<TableOfContents
			sections={[{ sections: [{ content: 'intro.md' }, { content: 'chapter.md' }] }]}
		/>
	);

	expect(actual.find('ComponentsList').prop('items')).toMatchInlineSnapshot(`
		Array [
		  Object {
		    "components": Array [],
		    "content": undefined,
		    "forcedOpen": false,
		    "heading": false,
		    "initialOpen": true,
		    "sections": Array [],
		    "selected": false,
		    "shouldOpenInNewTab": false,
		  },
		  Object {
		    "components": Array [],
		    "content": undefined,
		    "forcedOpen": false,
		    "heading": false,
		    "initialOpen": true,
		    "sections": Array [],
		    "selected": false,
		    "shouldOpenInNewTab": false,
		  },
		]
	`);
});

it('should render components of a single top section as root', () => {
	const actual = shallow(<TableOfContents sections={[{ components }]} />);

	expect(actual.find('ComponentsList').prop('items')).toMatchInlineSnapshot(`
		Array [
		  Object {
		    "components": Array [],
		    "content": undefined,
		    "forcedOpen": false,
		    "heading": false,
		    "initialOpen": true,
		    "name": "Button",
		    "sections": Array [],
		    "selected": false,
		    "shouldOpenInNewTab": false,
		    "slug": "button",
		  },
		  Object {
		    "components": Array [],
		    "content": undefined,
		    "forcedOpen": false,
		    "heading": false,
		    "initialOpen": true,
		    "name": "Input",
		    "sections": Array [],
		    "selected": false,
		    "shouldOpenInNewTab": false,
		    "slug": "input",
		  },
		  Object {
		    "components": Array [],
		    "content": undefined,
		    "forcedOpen": false,
		    "heading": false,
		    "initialOpen": true,
		    "name": "Textarea",
		    "sections": Array [],
		    "selected": false,
		    "shouldOpenInNewTab": false,
		    "slug": "textarea",
		  },
		]
	`);
});

/**
 * testing this layer with no mocking makes no sense...
 */
it('should render components with useRouterLinks', () => {
	const { getAllByRole } = render(
		<TableOfContents
			useRouterLinks
			sections={[
				{
					sections: [
						{ visibleName: '1', name: 'Components', slug: 'Components', content: 'intro.md' },
						{ visibleName: '2', content: 'chapter.md', slug: 'chap' },
					],
				},
			]}
		/>
	);

	expect((getAllByRole('link')[0] as any).href).toMatch(/\/#\/Components$/);
});

/**
 * testing this layer with no mocking makes no sense...
 * This test should not exist but for good coverage policy this is necessary
 */
it('should detect sections containing current selection when tocMode is collapse', () => {
	const context = {
		config: {
			tocMode: 'collapse',
		},
	};

	const Provider = (props: any) => <Context.Provider value={context} {...props} />;

	const { getByText } = render(
		<Provider>
			<TableOfContents
				tocMode="collapse"
				sections={[
					{
						sections: [
							{
								visibleName: '1',
								slug: 'Components',
								sections: [{ visibleName: '1.1', slug: 'Button' }],
							},
							{
								visibleName: '2',
								slug: 'chap',
								content: 'chapter.md',
								sections: [{ visibleName: '2.1', slug: 'Chapter #1' }],
							},
							{ visibleName: '3', href: 'http://react-styleguidist.com' },
						],
					},
				]}
				loc={{ pathname: '', hash: '/#Button' }}
			/>
		</Provider>
	);

	expect(getByText('1.1')).not.toBeEmpty();
});
