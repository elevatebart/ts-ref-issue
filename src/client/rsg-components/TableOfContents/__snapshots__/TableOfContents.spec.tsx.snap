// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render table of contents 1`] = `
<div>
  <div>
    <nav>
      <div>
        <input
          aria-label="Filter by name"
          onChange={[Function]}
          placeholder="Filter by name"
          value="foo"
        />
      </div>
      <div>
        foo
      </div>
    </nav>
  </div>
</div>
`;

exports[`should filter list when search field contains a query 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Button",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "button",
        },
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Input",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "input",
        },
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Textarea",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "textarea",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter list when search field contains a query 2`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm="but"
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": true,
          "heading": false,
          "initialOpen": true,
          "name": "Button",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "button",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter section names 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Introduction",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "introduction",
        },
        Object {
          "components": Array [
            Object {
              "name": "Button",
              "slug": "button",
            },
          ],
          "content": <ComponentsList
            hashPath={
              Array [
                "Buttons",
              ]
            }
            items={
              Array [
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Button",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "button",
                },
              ]
            }
            useHashId={true}
            useRouterLinks={false}
          />,
          "forcedOpen": false,
          "heading": true,
          "initialOpen": true,
          "name": "Buttons",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "buttons",
        },
        Object {
          "components": Array [
            Object {
              "name": "Input",
              "slug": "input",
            },
            Object {
              "name": "Textarea",
              "slug": "textarea",
            },
          ],
          "content": <ComponentsList
            hashPath={
              Array [
                "Forms",
              ]
            }
            items={
              Array [
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Input",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "input",
                },
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Textarea",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "textarea",
                },
              ]
            }
            useHashId={true}
            useRouterLinks={false}
          />,
          "forcedOpen": false,
          "heading": true,
          "initialOpen": true,
          "name": "Forms",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "forms",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter section names 2`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm="frm"
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": true,
          "heading": false,
          "initialOpen": true,
          "name": "Forms",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "forms",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should render a renderer 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Button",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "button",
        },
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Input",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "input",
        },
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Textarea",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "textarea",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should render renderer with sections with nested components 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    hashPath={Array []}
    items={
      Array [
        Object {
          "components": Array [],
          "content": undefined,
          "forcedOpen": false,
          "heading": false,
          "initialOpen": true,
          "name": "Introduction",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "introduction",
        },
        Object {
          "components": Array [
            Object {
              "name": "Button",
              "slug": "button",
            },
          ],
          "content": <ComponentsList
            hashPath={
              Array [
                "Buttons",
              ]
            }
            items={
              Array [
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Button",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "button",
                },
              ]
            }
            useHashId={true}
            useRouterLinks={false}
          />,
          "forcedOpen": false,
          "heading": true,
          "initialOpen": true,
          "name": "Buttons",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "buttons",
        },
        Object {
          "components": Array [
            Object {
              "name": "Input",
              "slug": "input",
            },
            Object {
              "name": "Textarea",
              "slug": "textarea",
            },
          ],
          "content": <ComponentsList
            hashPath={
              Array [
                "Forms",
              ]
            }
            items={
              Array [
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Input",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "input",
                },
                Object {
                  "content": undefined,
                  "forcedOpen": false,
                  "heading": false,
                  "initialOpen": true,
                  "name": "Textarea",
                  "selected": false,
                  "shouldOpenInNewTab": false,
                  "slug": "textarea",
                },
              ]
            }
            useHashId={true}
            useRouterLinks={false}
          />,
          "forcedOpen": false,
          "heading": true,
          "initialOpen": true,
          "name": "Forms",
          "sections": Array [],
          "selected": false,
          "shouldOpenInNewTab": false,
          "slug": "forms",
        },
      ]
    }
    useHashId={false}
    useRouterLinks={false}
  />
</Styled(TableOfContents)>
`;
